### PR TITLE
Renewal reminder emails and multiple reminders. #984, #558

### PR DESCRIPTION
**Pull request by nosegraze**
#984, #558

### CODE REVIEW

File: includes/admin/reminders/subscription-reminders.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	if ( ! is_admin() ) {
+		return;
+	}
+
+	if ( ! isset( $_GET['rcp-action'] ) || 'send_test_reminder' != $_GET['rcp-action'] ) {
+		return;
+	}
+
+	if ( ! current_user_can( 'rcp_manage_settings' ) ) {
+		wp_die( __( 'You do not have permission to delete reminder notices', 'rcp' ), __( 'Error', 'rcp' ), array( 'response' => 403 ) );
```
Comment (mindctrl) [inline includes/admin/reminders/subscription-reminders.php]: Wrong error message here.

File: includes/class-rcp-reminders.php [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+			'-2days'   => __( 'Two days after expiration', 'rcp' ),
+			'-3days'   => __( 'Three days after expiration', 'rcp' ),
+			'-1week'   => __( 'One week after expiration', 'rcp' ),
+			'-2weeks'  => __( 'Two weeks after expiration', 'rcp' ),
+			'-1month'  => __( 'One month after expiration', 'rcp' ),
+			'-2months' => __( 'Two months after expiration', 'rcp' ),
+			'-3months' => __( 'Three months after expiration', 'rcp' )
+		);
+
+		return apply_filters( 'rcp_reminder_notice_periods', $periods );
```
Comment (mindctrl) [inline includes/class-rcp-reminders.php]: Missed this filter.

File: includes/class-rcp-reminders.php [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+					$sent_time = get_user_meta( $member->ID, sanitize_key( '_rcp_reminder_sent_' . $member->get_subscription_id() . '_' . $notice_id ), true );
+
+					if ( $sent_time ) {
+						continue;
+					}
+
+					$rcp_email->member_id = $member->ID;
+					$rcp_email->send( $member->user_email, $notice['subject'], $notice['message'] );
+
+					rcp_add_member_note( $member->ID, sprintf( __( '%s notice was emailed to the member - %s.', 'rcp' ), ucwords( $type ), $this->get_notice_period_label( $notice_id ) ) );
```
Comment (mindctrl) [inline includes/class-rcp-reminders.php]: We can use `$member->add_note` here.

File: includes/class-rcp-reminders.php [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+	}
+
+	/**
+	 * Returns an array of reminder types and whether or not they're enabled
+	 *
+	 * @access public
+	 * @since  2.9
+	 * @return array Array of all notice types and whether or not they're enabled (true or false).
+	 */
+	public function get_enabled_reminders() {
```
Comment (mindctrl) [inline includes/class-rcp-reminders.php]: This should be named `get_enabled_reminder_types()`

File: includes/class-rcp-reminders.php [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+				$args['meta_query'][] = array(
+					'key'     => 'rcp_status',
+					'compare' => 'IN',
+					'value'   => $status
+				);
+				break;
+
+		}
+
+		$args = apply_filters( 'rcp_reminder_subscription_args', $args, $period, $type );
```
Comment (mindctrl) [inline includes/class-rcp-reminders.php]: I missed this filter earlier. We need to remove it too for now.

**Comment by pippinsplugins** [comment]
@nosegraze looks like we have a merge conflict now.

**Comment by nosegraze** [comment]
Fixed!

File: includes/admin/reminders/subscription-reminder-view.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+					<input type="checkbox" id="rcp-notice-enabled" name="enabled" value="1" <?php checked( $notice['enabled'], true ); ?>>
+					<span><?php _e( 'Check to enable sending the email. If unchecked, the email won\'t be sent to members.', 'rcp' ); ?></span>
+				</td>
+			</tr>
+			<tr>
+				<th scope="row" valign="top">
+					<label for="rcp-notice-message"><?php _e( 'Email Message', 'rcp' ); ?></label>
+				</th>
+				<td>
+					<?php wp_editor( wpautop( wp_kses_post( wptexturize( $notice['message'] ) ) ), 'message', array( 'textarea_name' => 'message' ) ); ?>
```
Comment (mindctrl) [inline includes/admin/reminders/subscription-reminder-view.php]: We should go ahead and prefix this textarea name to make it less generic.

File: includes/admin/reminders/subscription-reminders.php [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	}
+
+	update_option( 'rcp_reminder_notices', $notices );
+
+	wp_safe_redirect( $redirect_url );
+	exit;
+
+}
+
+add_action( 'admin_init', 'rcp_process_add_edit_reminder_notice' );
```
Comment (mindctrl) [inline includes/admin/reminders/subscription-reminders.php]: We can move this to the new rcp_action processing and avoid running it on admin_init all the time. Same with the `rcp_process_delete_reminder_notice` and `rcp_process_send_test_reminder_notice`

**Pull request merged**